upload.py

```diff
@@ -94,89 +97,152 @@ def GatherPreviousDownloads(connection, trailing_days):
-
-         if (downloads > 10) and (previous.downloads * 115 // 100 < downloads):
-           print(sample.product, sample.file, sample.version, sample.sample_date,
-                 downloads, 'large jump from %d' % previous.downloads)
-
-      cmd = """INSERT INTO gh_downloads(
-          sample_date, filename, downloads_total, sha256_total, sig_total,
-          product, version, arch, os, extension, is_installer,
-          downloads, sha256, sig)
-      VALUES(
-          '%s', '%s', %d, %d, %d, '%s', '%s', '%s', '%s', '%s', %d,
-          %d, %d, %d
-      )""" % (sample.sample_date, sample.file,
-              sample.downloads_total, sample.sha256_total, sample.sig_total,
-              sample.product, sample.version, sample.arch, sample.os,
-              sample.extension, 1 if sample.installer else 0,
-              downloads, sha256, sig,
-              )
-      history[(sample.product, sample.file, sample.version, day)] = \
-          DownloadSample(
-              file=sample.file,
-              sample_date=sample.sample_date,
-              downloads_total=sample.downloads_total,
-              sha256_total=sample.sha256_total,
-              sig_total=sample.sig_total,
-              product=sample.product,
-              version=sample.version,
-              arch=sample.arch,
-              os=sample.os,
-              extension=sample.extension,
-              installer=sample.installer,
-              downloads=downloads,
-              sha256=sha256,
-              sig=sig)
-
-      if not dry_run:
-        cursor.execute(cmd)
-  if not dry_run:
-    cursor.close()
-    connection.commit()
+class DailyCountUploader(object):
+
+  def __init__(self, history, connection, window=1, backfill=True,
+               dry_run=True):
+    self.history = history
+    self.connection = connection
+    self.window = window
+    self.backfill = backfill
+    self.dry_run = dry_run
+    self.cursor = None
+
+  def upload_file(self, file):
+    if not self.dry_run:
+      self.cursor = self.connection.cursor()
+
+    with open(file, 'r') as inp:
+      print('uploading:', file)
+      for line in inp:
+        # file| ymd | hm | count | #sha | #sig | product | version | arch | os
+        #     extension
+        parts = line.strip().split('|')
+        sample = DownloadSample(
+            file=parts[0],
+            sample_date=parts[1],
+            downloads_total=int(parts[3]),
+            sha256_total=int(parts[4]),
+            sig_total=int(parts[5]),
+            product=parts[6],
+            version=none_to_null(parts[7]),
+            arch=none_to_null(parts[8]),
+            os=none_to_null(parts[9]),
+            extension=none_to_null(parts[10]),
+            installer=parts[11] == 'installer',
+            downloads=0,
+            sha256=0,
+            sig=0)
+        self.process_sample(sample)
+
+    if not self.dry_run:
+      self.cursor.close()
+      self.connection.commit()
+
+  def process_sample(self, sample):
```

Comment:
please write a comment on what this function is processing.

Reply:
done